### PR TITLE
FEAT: Edb cfg export cutout

### PR DESCRIPTION
**Pull request by ring630**
- export cutout to json
- new method edb.layout.find_primitive
- separate test for edb.layout

### CODE REVIEW

File: src/pyedb/dotnet/edb_core/cell/layout.py [hunk 1]
```diff
@@ -331,3 +331,23 @@ def find_component_by_name(self, value: str):
...
+    def find_primitive(self, layer_name: str = None):
+        """Find a primitive objects by layer name.
+
+        Parameters
+        ----------
+        layer_name : str
+            Name of the layer.
+        Returns
+        -------
+
```
Comment (svandenb-dev) [inline src/pyedb/dotnet/edb_core/cell/layout.py:344]: missing returned object in doc strings

File: src/pyedb/dotnet/edb_core/cell/layout.py [hunk 2]
```diff
@@ -331,3 +331,23 @@ def find_component_by_name(self, value: str):
...
+        """
+        all_prims = self.primitives
+        prims = []
```
Comment (svandenb-dev) [inline src/pyedb/dotnet/edb_core/cell/layout.py:347]:
```suggestion
        
```

File: src/pyedb/dotnet/edb_core/cell/layout.py [hunk 3]
```diff
@@ -331,3 +331,23 @@ def find_component_by_name(self, value: str):
...
+        if layer_name:
+            for i in all_prims:
```
Comment (svandenb-dev) [inline src/pyedb/dotnet/edb_core/cell/layout.py:349]:
```suggestion
            return [prim for prim in self.primtives if prim.layer_name == layer_name]
```

File: src/pyedb/dotnet/edb_core/cell/layout.py [hunk 4]
```diff
@@ -331,3 +331,23 @@ def find_component_by_name(self, value: str):
...
+                if i.layer_name == layer_name:
```
Comment (svandenb-dev) [inline src/pyedb/dotnet/edb_core/cell/layout.py:350]:
```suggestion
                
```

File: src/pyedb/dotnet/edb_core/cell/layout.py [hunk 5]
```diff
@@ -331,3 +331,23 @@ def find_component_by_name(self, value: str):
...
+                    prims.append(i)
```
Comment (svandenb-dev) [inline src/pyedb/dotnet/edb_core/cell/layout.py:351]:
```suggestion
                    
```

File: src/pyedb/dotnet/edb_core/cell/layout.py [hunk 6]
```diff
@@ -331,3 +331,23 @@ def find_component_by_name(self, value: str):
...
+            all_prims = prims
```
Comment (svandenb-dev) [inline src/pyedb/dotnet/edb_core/cell/layout.py:352]:
```suggestion
            
```

File: src/pyedb/dotnet/edb_core/cell/layout.py [hunk 7]
```diff
@@ -331,3 +331,23 @@ def find_component_by_name(self, value: str):
...
+        return all_prims
```
Comment (svandenb-dev) [inline src/pyedb/dotnet/edb_core/cell/layout.py:353]:
```suggestion
        
```

File: src/pyedb/dotnet/edb_core/cell/layout.py [hunk 8]
```diff
@@ -331,3 +331,23 @@ def find_component_by_name(self, value: str):
...
+
+        Parameters
+        ----------
+        layer_name : str
+            Name of the layer.
+        Returns
+        -------
+
+        """
+        all_prims = self.primitives
```
Comment (svandenb-dev) [inline src/pyedb/dotnet/edb_core/cell/layout.py:346]:
```suggestion
        
```

**Review by svandenb-dev — DISMISSED** [review]
LGTM

**Comment by ring630** [comment]
@svandenb-dev I updated the code according to your comments. Please add your comments as one review comment next time.

File: src/pyedb/dotnet/edb_core/cell/layout.py [hunk 9]
```diff
@@ -331,3 +333,21 @@ def find_component_by_name(self, value: str):
         """
         obj = self._pedb._edb.Cell.Hierarchy.Component.FindByName(self._edb_object, value)
         return EDBComponent(self._pedb, obj) if obj is not None else None
+
+    def find_primitive(self, layer_name: Union[str, list]) -> list:
```
Comment (svandenb-dev) [inline src/pyedb/dotnet/edb_core/cell/layout.py:337]: Union[str, list] is a great feature but I think is not supported in python 3.7 I had to remove last time as it was failing for users. I would check.

File: src/pyedb/dotnet/edb_core/cell/layout.py [hunk 10]
```diff
@@ -331,3 +333,21 @@ def find_component_by_name(self, value: str):
...
+        Returns
+        -------
+        list
+        """
+        layer_name = layer_name if isinstance(layer_name, list) else [layer_name]
+        prims = []
+        for i in self.primitives:
+            if i.layer_name in layer_name:
+                prims.append(i)
+        return prims
```
Comment (svandenb-dev) [inline src/pyedb/dotnet/edb_core/cell/layout.py:353]: I would use list comprehension.
return [prim for prim in self.primitives if prim.layer_name == layer_name]

**Review by svandenb-dev — DISMISSED** [review]
LGTM

File: src/pyedb/dotnet/edb_core/cell/layout.py (hunk 9 above)
Comment (ring630) [reply src/pyedb/dotnet/edb_core/cell/layout.py:337]: It's been used in places already. Does pyedbg still support python3.7

Comment (svandenb-dev) [reply src/pyedb/dotnet/edb_core/cell/layout.py:337]: Yes for few months

Comment (svandenb-dev) [reply src/pyedb/dotnet/edb_core/cell/layout.py:337]: checked work with python 3.8 closing.

**Pull request merged**